### PR TITLE
[Storage] Fix#10633: Support Azure Files AD DS Authentication

### LINKED ISSUE #10633: Azure Files AD DS Authentication
**Resource Provider**
What is the Azure resource provider your feature is part of?
Azure Storage

**Description of Feature or Work Requested**
We would like to release a Preview Module for Azure Files native AD Authentication in Mid Oct. Zhenfeng from SRP team is working on the SRP changes and the changes including swaggers will be available by 9/13. We would need your help to assess the dev cost and priority for the next CLI release. I have also attached the discussion thread we had with Wei on PSH. Unlike PSH, we are only extending the existing commands without introducing new ones.
 
1.	Extend az storage account keys list and az storage account keys renew to support two new Kerb keys
2.	Support new flag --enable-files-adds for az storage account create and update
You can find the feature enablement matrix similar as what is proposed for PSH

Initial DirectoryServiceOptions of Account | None | AADDS | AD
-- | -- | -- | --
Set   Account with “–EnableAzureActiveDirectoryDomainServicesForFile $false”, | None | None | AD
Set   Account with “–EnableAzureActiveDirectoryDomainServicesForFile $true”, | AADDS | AADDS | Blocked   with error msg above
Set   Account with “–EnableActiveDirectoryDomainServicesForFile $false”, | None | AADDS | None
Set   Account with “–EnableActiveDirectoryDomainServicesForFile $true”, | AD | Blocked   with error msg above | AD

**Minimum API Version Required**
What is the minimum API version of your service required to implement your feature?

**Swagger Link**
Provide a link to the location of your feature(s) in the REST API specs repo. If your feature(s) has corresponding commit or pull request in the REST API specs repo, provide them. This should be on the master branch of the REST API specs repo.

**Target Date**
Before Ignite Nov 4th

### PR DESCRIPTION
**Pull request by Juliehzl**
Fix #10633

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

File: src/azure-cli/azure/cli/command_modules/storage/_params.py [hunk 1]
```diff
@@ -72,7 +72,29 @@ def load_arguments(self, _):  # pylint: disable=too-many-locals, too-many-statem
              'Once the property is enabled, the feature cannot be disabled. Currently only supported for LRS and '
              'ZRS replication types, hence account conversions to geo-redundant accounts would not be possible. '
              'For more information, please refer to https://go.microsoft.com/fwlink/?linkid=2086047.')
-
+    adds_type = CLIArgumentType(arg_type=get_three_state_flag(), min_api='2019-04-01',
+                                help='Enable Azure Files Active Directory Domain Service Authentication for '
+                                     'storage account. When --enable-files-adds is set to true, Azure Active '
+                                     'Directory Properties arguments must be provided.')
+    aadds_type = CLIArgumentType(arg_type=get_three_state_flag(), min_api='2018-11-01',
```
Comment (MyronFanQiu) [inline src/azure-cli/azure/cli/command_modules/storage/_params.py]: should we add `arg_group = "Azure Active Directory Properties`? Just confirm

File: src/azure-cli/azure/cli/command_modules/storage/_params.py [hunk 2]
```diff
@@ -169,9 +203,16 @@ def load_arguments(self, _):  # pylint: disable=too-many-locals, too-many-statem
...
-        c.argument('key_name', options_list=['--key'], help='The key to regenerate.', validator=validate_key,
+    with self.argument_context('storage account keys list', resource_type=ResourceType.MGMT_STORAGE) as c:
+        t_expand_key_type = self.get_models('ListKeyExpand', resource_type=ResourceType.MGMT_STORAGE)
+        c.argument("expand", options_list=['--expand-key-type'], help='Specify the expanded key types to be listed.',
+                   arg_type=get_enum_type(t_expand_key_type), min_api='2019-04-01', is_preview=True)
+
+    with self.argument_context('storage account keys renew', resource_type=ResourceType.MGMT_STORAGE) as c:
+        c.argument('key_name', options_list=['--key'], help='The key options to regenerate.',
                    arg_type=get_enum_type(list(storage_account_key_options.keys())))
+        c.extra('key_type', help='The key type to regenerate. If --key-type is not specified, access keys will be '
```
Comment (MyronFanQiu) [inline src/azure-cli/azure/cli/command_modules/storage/_params.py]: only one candidate?

Comment (MyronFanQiu) [reply src/azure-cli/azure/cli/command_modules/storage/_params.py]: regenerated by default

File: src/azure-cli/azure/cli/command_modules/storage/operations/account.py [hunk 3]
```diff
@@ -35,10 +37,40 @@ def create_storage_account(cmd, resource_group_name, account_name, sku=None, loc
...
+    AzureFilesIdentityBasedAuthentication = cmd.get_models('AzureFilesIdentityBasedAuthentication')
     if enable_files_aadds is not None:
-        AzureFilesIdentityBasedAuthentication = cmd.get_models('AzureFilesIdentityBasedAuthentication')
         params.azure_files_identity_based_authentication = AzureFilesIdentityBasedAuthentication(
             directory_service_options='AADDS' if enable_files_aadds else 'None')
+    if enable_files_adds is not None:
+        from knack.util import CLIError
+        ActiveDirectoryProperties = cmd.get_models('ActiveDirectoryProperties')
+        if enable_files_adds:  # enable AD
+            if not (domain_name and net_bios_domain_name and forest_name and domain_guid and domain_sid and
```
Comment (MyronFanQiu) [inline src/azure-cli/azure/cli/command_modules/storage/operations/account.py]: is it better to write a validator ?

File: src/azure-cli/azure/cli/command_modules/storage/operations/account.py [hunk 4]
```diff
@@ -144,10 +178,62 @@ def update_storage_account(cmd, instance, sku=None, tags=None, custom_domain=Non
...
-        params.azure_files_identity_based_authentication = AzureFilesIdentityBasedAuthentication(
-            directory_service_options='AADDS' if enable_files_aadds else 'None')
+        if enable_files_aadds:  # enable AADDS
+            params.azure_files_identity_based_authentication = AzureFilesIdentityBasedAuthentication(
+                directory_service_options='AADDS' if enable_files_aadds else 'None')
+        else:  # Only disable AADDS and keep others unchanged
+            scf = storage_client_factory(cmd.cli_ctx)
+            from msrestazure.tools import parse_resource_id
+            rg = parse_resource_id(instance.id)['resource_group']
+            origin_storage_account = scf.storage_accounts.get_properties(rg, instance.name)
```
Comment (MyronFanQiu) [inline src/azure-cli/azure/cli/command_modules/storage/operations/account.py]: a weird behavior for service....

File: src/azure-cli/azure/cli/command_modules/storage/operations/account.py [hunk 5]
```diff
@@ -231,3 +317,9 @@ def update_management_policies(client, resource_group_name, account_name, parame
     if parameters:
         parameters = parameters.policy
     return client.create_or_update(resource_group_name, account_name, policy=parameters)
+
+
+def list_expand(client, resource_group_name, account_name, parameters=None):
```
Comment (MyronFanQiu) [inline src/azure-cli/azure/cli/command_modules/storage/operations/account.py]: what's this used for?

File: src/azure-cli/azure/cli/command_modules/storage/operations/account.py (hunk 4 above)
Comment (MyronFanQiu) [reply src/azure-cli/azure/cli/command_modules/storage/operations/account.py]: an extra util function may be helpful

File: src/azure-cli/azure/cli/command_modules/storage/_params.py (hunk 1 above)
Comment (Juliehzl) [reply src/azure-cli/azure/cli/command_modules/storage/_params.py]: For aadds, no azure active directory properties

File: src/azure-cli/azure/cli/command_modules/storage/_params.py (hunk 2 above)
Comment (Juliehzl) [reply src/azure-cli/azure/cli/command_modules/storage/_params.py]: > only one candidate?

Yes

File: src/azure-cli/azure/cli/command_modules/storage/operations/account.py (hunk 5 above)
Comment (Juliehzl) [reply src/azure-cli/azure/cli/command_modules/storage/operations/account.py]: deleted

File: src/azure-cli/HISTORY.rst [hunk 6]
```diff
@@ -84,6 +84,11 @@ Release History
 
 * Add "--compute-model", "--auto-pause-delay", and "--min-capacity" parameters to support CRUD operations for new SQL Database offering: Serverless compute model."
 
+**Storage**
+
+* storage account create/update: Add --enable-files-adds parameter and Azure Active Directory Properties Argument group to support Azure Files Active Directory Domain Service Authentication
+* Expand `az storage account keys list/renew` to support list or regenerate Kerberos keys of Storage account.
```
Comment (qwordy) [inline src/azure-cli/HISTORY.rst]: support listing or regenerating keys

**Pull request merged**